### ISSUE TITLE
Ansible is getting host IPs from where it shouldn't

### ISSUE THREAD

**Issue by eregimbald** [ansible/ansible#23465]
##### ISSUE TYPE
Bug Report

##### COMPONENT NAME
Shell

##### ANSIBLE VERSION
ansible 2.2.1.0

##### CONFIGURATION
default

##### OS / ENVIRONMENT
Distributor ID: Ubuntu
Description:    Ubuntu 16.04.1 LTS
Release:        16.04
Codename:       xenial

##### SUMMARY
I'm running a playbook on an empty host list and Ansible grabs the IP of a switch somewhere on our network

##### STEPS TO REPRODUCE
vars:
  mylist:[]
hosts: mylist

run the playbook

```
---
- name: Deploy a unique configuration
  vars:
    hostlist: []
  hosts: "{{hostlist}}"
  gather_facts: no
  connection: local
  
  pre_tasks:
    - shell: ls -1
      args:
        chdir: "{{ playbook_dir }}/vars/deploy-config/"
      register: list1
    - set_fact: 
        hostlist="{{list1.stdout_lines}}"
    - debug: var=hostlist
    
  roles:
    - {role: cleanup-create, commit: 0}
```

This is the result of running the playbook:

```
$ ansible-playbook deploy-configuration.yml

PLAY [Deploy a unique configuration] *******************************************

TASK [command] *****************************************************************
changed: [10.200.254.3]

TASK [set_fact] ****************************************************************
ok: [10.200.254.3]

TASK [debug] *******************************************************************
ok: [10.200.254.3] => {
    "hostlist": [
        "testfile1",
        "testfile2",
        "testfile3"
    ]
}

TASK [cleanup-create : Removing the old config directories] ********************
skipping: [10.200.254.3]

TASK [cleanup-create : Creating the candidate config directories] **************
skipping: [10.200.254.3]

PLAY RECAP *********************************************************************
10.200.254.3               : ok=3    changed=1    unreachable=0    failed=0

```

##### EXPECTED RESULTS
I was trying to get a list of hosts equal to the names of all the files in my folder, however Ansible evaluates hosts before anything so I don't think this is possible. In any case I don't see how it was able to learn the IP 10.200.254.3. It's the lo0 interface on a switch, somewhere.

##### ACTUAL RESULTS

TASK [command] *****************************************************************
task path: /home/eregimbald/network-ansible-eregimbald/deploy-configuration.yml:10
Using module file /usr/local/lib/python2.7/dist-packages/ansible/modules/core/commands/command.py
<10.200.254.3> ESTABLISH LOCAL CONNECTION FOR USER: eregimbald
<10.200.254.3> EXEC /bin/sh -c '( umask 77 && mkdir -p "` echo ~/.ansible/tmp/ansible-tmp-1491839004.99-68109784129678 `" && echo ansible-tmp-1491839004.99-68109784129678="` echo ~/.ansible/tmp/ansible-tmp-1491839004.99-68109784129678 `" ) && sleep 0'
<10.200.254.3> PUT /tmp/tmpFHi5J3 TO /home/eregimbald/.ansible/tmp/ansible-tmp-1491839004.99-68109784129678/command.py
<10.200.254.3> EXEC /bin/sh -c 'chmod u+x /home/eregimbald/.ansible/tmp/ansible-tmp-1491839004.99-68109784129678/ /home/eregimbald/.ansible/tmp/ansible-tmp-1491839004.99-68109784129678/command.py && sleep 0'
<10.200.254.3> EXEC /bin/sh -c '/usr/bin/python /home/eregimbald/.ansible/tmp/ansible-tmp-1491839004.99-68109784129678/command.py; rm -rf "/home/eregimbald/.ansible/tmp/ansible-tmp-1491839004.99-68109784129678/" > /dev/null 2>&1 && sleep 0'

```

```

**Comment by abadger** [comment]
Indeed, getting host list that way won't work.

I wasn't able to reproduce the behaviour of ansible with the playbook you gave but I had to remove the role since I don't have it.  (Tried with both a checkout of 2.2.1.0 and devel) My output is:
```
$ ansible-playbook 23465.yml                                     (13:06:30)
 [WARNING]: Host file not found: /etc/ansible/hosts

 [WARNING]: provided hosts list is empty, only localhost is available

 [WARNING]: Could not match supplied host pattern, ignoring: {{hostlist}}


PLAY [Deploy a unique configuration] ********************************************************************************
skipping: no hosts matched

PLAY RECAP **********************************************************************************************************
```

Are you able to reduce it down to a minimum amount that you can reproduce with?

needs_info

**Comment by eregimbald** [comment]
I found out what happens. 

It seems that if you configure hosts with an empty variable, Ansible will default to the contents of its inventory. When I ran the book with -i myinventory.hosts, it would run the book on every host in that inventory file while if I ran it without -i it would run on that switch.

Somehow, that IP address ended up in /etc/ansible/hosts. I'm not sure what it's doing there since I never use the default inventory.

TIL...
